kusto/kcsb.go

```diff
@@ -0,0 +1,362 @@
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Interactive Login. Error: %s", err)
+		}
+		tkp.tokenCred = cred
+	} else if !isEmpty(kcsb.AADUserID) && !isEmpty(kcsb.Password) {
+		var ops *azidentity.UsernamePasswordCredentialOptions
+
+		if kcsb.ClientOptions != nil {
+			ops = &azidentity.UsernamePasswordCredentialOptions{ClientOptions: *kcsb.ClientOptions}
+		}
+
+		cred, err := azidentity.NewUsernamePasswordCredential(kcsb.AuthorityId, kcsb.ApplicationClientId, kcsb.AADUserID, kcsb.Password, ops)
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Username Password. Error: %s", err)
+		}
+		tkp.tokenCred = cred
+	} else if !isEmpty(kcsb.ApplicationClientId) && !isEmpty(kcsb.ApplicationKey) {
+		cred, err := azidentity.NewClientSecretCredential(kcsb.AuthorityId, kcsb.ApplicationClientId, kcsb.ApplicationKey, &azidentity.ClientSecretCredentialOptions{ClientOptions: *kcsb.ClientOptions})
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Client Secret. Error: %s", err)
+		}
+		tkp.tokenCred = cred
+		return tkp, nil
+	} else if !isEmpty(kcsb.ApplicationCertificate) {
+		cert, thumprintKey, err := azidentity.ParseCertificates([]byte(kcsb.ApplicationCertificate), []byte(kcsb.ApplicationCertificateThumbprint))
+		if err != nil {
+			return nil, err
+		}
+		opts := &azidentity.ClientCertificateCredentialOptions{}
+		opts.SendCertificateChain = kcsb.SendCertificateChain
+		if kcsb.ClientOptions != nil {
+			opts.ClientOptions = *kcsb.ClientOptions
+		}
+		cred, err := azidentity.NewClientCertificateCredential(kcsb.AuthorityId, kcsb.ApplicationClientId, cert, thumprintKey, opts)
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Application Certificate: %s", err)
+		}
+		tkp.tokenCred = cred
+	} else if kcsb.MSIAuthentication {
+		opts := &azidentity.ManagedIdentityCredentialOptions{}
+		if !isEmpty(kcsb.ManagedServiceIdentity) {
+			opts.ID = azidentity.ClientID(kcsb.ManagedServiceIdentity)
+		}
+		if kcsb.ClientOptions != nil {
+			opts.ClientOptions = *kcsb.ClientOptions
+		}
+
+		cred, err := azidentity.NewManagedIdentityCredential(opts)
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Managed Identity: %s", err)
+		}
+		tkp.tokenCred = cred
+	} else if !isEmpty(kcsb.UserToken) {
+		tkp.customToken = kcsb.UserToken
+	} else if !isEmpty(kcsb.ApplicationToken) {
+		tkp.customToken = kcsb.ApplicationToken
+	} else if kcsb.AZCLI {
+		opts := &azidentity.AzureCLICredentialOptions{}
+		opts.TenantID = kcsb.AuthorityId
+		cred, err := azidentity.NewAzureCLICredential(opts)
+		if err != nil {
+			return nil, fmt.Errorf("Error : Couldn't retrieve client credentiels using Azure CLI: %s", err)
+		}
+		tkp.tokenCred = cred
+	} else {
+		//env variables based auth
+		opts := &azidentity.EnvironmentCredentialOptions{}
```

Comment:
No way for the user to provide a kusto credential?

Reply:
As per discussion with Yochai, it seems the Kusto credential is just used for C# code, that too internal process. It doesn't make sense to be present in SDKs